### PR TITLE
Be compatible with all in-use cppzmq versions.

### LINKED ISSUE #273: Can not compile libtango on Ubuntu 16 using libzmq3-dev package from apt
Error message

```
[ 17%] Building CXX object cppapi/client/CMakeFiles/client_objects.dir/zmqeventconsumer.cpp.o
/home/ingvord/Projects/org.tango/git/cppTango/cppapi/client/zmqeventconsumer.cpp: In member function ‘virtual void* Tango::ZmqEventConsumer::run_undetached(void*)’:
/home/ingvord/Projects/org.tango/git/cppTango/cppapi/client/zmqeventconsumer.cpp:186:18: error: cannot convert ‘zmq::socket_t’ to ‘void*’ in assignment
  items[0].socket = *control_sock;
                  ^
/home/ingvord/Projects/org.tango/git/cppTango/cppapi/client/zmqeventconsumer.cpp:187:18: error: cannot convert ‘zmq::socket_t’ to ‘void*’ in assignment
  items[1].socket = *heartbeat_sub_sock;
                  ^
/home/ingvord/Projects/org.tango/git/cppTango/cppapi/client/zmqeventconsumer.cpp:188:18: error: cannot convert ‘zmq::socket_t’ to ‘void*’ in assignment
  items[2].socket = *event_sub_sock;
                  ^
/home/ingvord/Projects/org.tango/git/cppTango/cppapi/client/zmqeventconsumer.cpp: In member function ‘bool Tango::ZmqEventConsumer::process_ctrl(zmq::message_t&, zmq::pollitem_t*, int&)’:
/home/ingvord/Projects/org.tango/git/cppTango/cppapi/client/zmqeventconsumer.cpp:1063:47: error: cannot convert ‘zmq::socket_t’ to ‘void*’ in assignment
                 poll_list[old_poll_nb].socket = *tmp_sock;

```

```
Linux ingvord-VirtualBox 4.4.0-38-generic #57-Ubuntu SMP Tue Sep 6 15:42:33 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```

libzmq3-dev 4.1.4-7

### LINKED ISSUE #499: compilation on ubuntu 18.04 with  gcc-7.3.0 and zeromq-4.2.5 fails
Compilation on ubuntu 18.04 with gcc-7.3.0 and zeromq-4.2.5 fails with the following error:
cppTango-9.3.2/cppapi/client/zmqeventconsumer.cpp:184:21: error: cannot convert ‘zmq::socket_t’ to ‘void*’ in assignment
  items[0].socket = *control_sock;
(the same error is reapetd for serveral times)

The same error occurs with zmq-4.2.2 from GitHub. Compiling with zmq-4.0.7 is ok.
Googling around it seems that the reported error si common for prjetcs using recnt zeromq and C++11.

By patching two files the compilation is successful (see attached patches). Tests of the reulting library are under way.
[zmq_socket_patch_1.txt](https://github.com/tango-controls/cppTango/files/2595954/zmq_socket_patch_1.txt)
[zmq_socket_patch_2.txt](https://github.com/tango-controls/cppTango/files/2595955/zmq_socket_patch_2.txt)

### LINKED ISSUE #535: Travis build fails on cppzmq for CentOS7
I'm trying to set up Travis build for CentOS 7, however I got stuck on the `install_cppzmq.sh` script.
It's giving me these errors:
```
$ .travis/install_cppzmq.sh
Build cppzmq
-- The C compiler identification is GNU 4.8.5
-- The CXX compiler identification is GNU 4.8.5
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Warning at CMakeLists.txt:4 (find_package):
  By not providing "FindZeroMQ.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "ZeroMQ", but
  CMake did not find one.
  Could not find a package configuration file provided by "ZeroMQ" with any
  of the following names:
    ZeroMQConfig.cmake
    zeromq-config.cmake
  Add the installation prefix of "ZeroMQ" to CMAKE_PREFIX_PATH or set
  "ZeroMQ_DIR" to a directory containing one of the above files.  If "ZeroMQ"
  provides a separate development package or SDK, be sure it has been
  installed.
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.27.1") 
CMake Error at CMakeLists.txt:12 (message):
  ZeroMQ was NOT found!
-- Configuring incomplete, errors occurred!
See also "/home/tango/cppzmq/build/CMakeFiles/CMakeOutput.log".
The command ".travis/install_cppzmq.sh" failed and exited with 255 during .
Your build has been stopped.
```
I can see it's missing some cmake files, but where should I get those from? And why it's working for Debian builds?

You can see the configuration I have right now [here](https://github.com/synchrotron-solaris/cppTango/tree/centos7_build) (`centos7_build` branch of synchrotron-solaris' fork).

### PR DESCRIPTION
**Pull request by t-b**
Follow up from #530.

Close #499.
Close #273.
Close #535.

### CODE REVIEW

**Comment by mliszcz** [comment]
Looks like there is a problem with cppzmq from debian 7 repos:

```
/home/tango/src/cppapi/client/zmqeventconsumer.cpp:3553:16: error: 'class zmq::socket_t' has no member named 'disconnect'
```

**Comment by t-b** [comment]
@mliszcz Okay so I'll add a test for disconnect and resurrect the manual cppzmq installation. Nobody should use debian7 (wheezy) anymore as it is EOL, but let's not get distracted here.

**Comment by mliszcz** [comment]
Why do you need a test for disconnect? Maybe just revert `install_cppzmq.sh` and use if only for debian 7.

**Comment by t-b** [comment]
@mliszcz The test is required so that people how are trying to compile it on debian 7 don't get a compile error but an error message from cmake.

File: .travis/debian7/Dockerfile [hunk 1]
```diff
@@ -10,7 +10,9 @@ RUN echo 'deb http://archive.debian.org/debian wheezy-backports main contrib' >
 
 RUN echo "Acquire::Check-Valid-Until false;" > /etc/apt/apt.conf
 
-RUN apt-get update && apt-get install -y apt-utils
+RUN apt-get update
```
Comment (mliszcz) [inline .travis/debian7/Dockerfile]: I understand that these images are thrown away after the CI job is completed, but this change goes directly against official docker "best practices". Be careful when applying it to other images.

https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#run

Comment (t-b) [reply .travis/debian7/Dockerfile]: Thanks for the hint.

**Comment by bourtemb** [comment]
I think the issue with zmq disconnect is actually due to a recent refactoring change when ZmqEventConsumer::disconnect_socket() method was created.
I think we should simply enclose the definition of ZmqEventConsumer::disconnect_socket() inside a
#ifdef ZMQ_HAS_DISCONNECT section and it should compile without having the need to do the check at cmake level.

**Comment by t-b** [comment]
@bourtemb The `disconnect` method was introduced in 2013 and only cppzmq of debian wheezy does not have that. I'll make the `#ifdef` changes you suggested.

**Comment by t-b** [comment]
@bourtemb Turns out we already had the ZMQ_HAS_DISCONNECT define. I've done as you suggested and also removed my try compile as we can how handle both cases properly.

**Pull request merged**